llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,461 @@
+/// could cause WAW issues in the loop pipelining.
+/// The pass will update the \p VirtRegMap so that the new vregs have fixed
+/// assignments, guaranteeing that the reg renaming does not cause conflicts.
+class AIEWawRegRewriter : public MachineFunctionPass {
+
+public:
+  static char ID;
+  AIEWawRegRewriter() : MachineFunctionPass(ID) {}
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addRequired<VirtRegMap>();
+    AU.addPreserved<VirtRegMap>();
+    AU.addRequired<SlotIndexes>();
+    AU.addPreserved<SlotIndexes>();
+    AU.addRequired<LiveIntervals>();
+    AU.addPreserved<LiveIntervals>();
+    AU.addRequired<LiveRegMatrix>();
+    AU.addPreserved<LiveRegMatrix>();
+    AU.addRequired<MachineLoopInfo>();
+    AU.addPreserved<MachineLoopInfo>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
+};
+
+static bool
+renameMBBPhysRegs(const SmallVector<const MachineBasicBlock *, 4> &LoopBBs,
+                  const MachineFunction &MF, VirtRegMap &VRM,
+                  LiveRegMatrix &LRM, const MachineRegisterInfo &MRI,
+                  const LiveIntervals &LIS, const AIEBaseRegisterInfo &TRI);
+
+/// Find Free register of the same register Class type and
+// exclude the forbidden Physical register from the result
+static SmallVector<MCPhysReg, 16> getPriorityReplacementRegs(
+    const Register Reg, const SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+    const MachineFunction &MF, const VirtRegMap &VRM, LiveRegMatrix &LRM,
+    const MachineRegisterInfo &MRI, const LiveIntervals &LIS,
+    const AIEBaseRegisterInfo &TRI);
+
+static bool replaceReg(const Register Reg,
+                       SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+                       const MachineFunction &MF, VirtRegMap &VRM,
+                       LiveRegMatrix &LRM, const MachineRegisterInfo &MRI,
+                       const LiveIntervals &LIS,
+                       const AIEBaseRegisterInfo &TRI);
+
+static bool isWAWCandidate(const Register &Reg, const MachineInstr &MI,
+                           SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+                           SmallVector<MCPhysReg, 16> &UsedVirtRegs,
+                           VirtRegMap &VRM, const MachineRegisterInfo &MRI);
+
+bool isIdentityCopy(const MachineInstr &MI, const VirtRegMap &VRM);
+
+static SmallSet<MCPhysReg, 16>
+getBBForbiddenReplacements(const MachineBasicBlock *MBB, const VirtRegMap &VRM,
+                           const MachineRegisterInfo &MRI,
+                           const AIEBaseRegisterInfo &TRI);
+
+static void addAliasRegs(SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+                         const MCPhysReg &PhysReg,
+                         const MachineRegisterInfo &MRI,
+                         const AIEBaseRegisterInfo &TRI);
+
+// Check if the UnDef is the last define for a Register, or if there is
```

Comment:
nit: ``UnDef ` -> `Def` .

Reply:
no this is actually a undef. 
A register can also be used, when it was only written to with a undef. A unit test clears this up